client/alloc_watcher.go

```diff
@@ -0,0 +1,448 @@
+package client
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/hashicorp/consul/lib"
+	nomadapi "github.com/hashicorp/nomad/api"
+	"github.com/hashicorp/nomad/client/allocdir"
+	"github.com/hashicorp/nomad/client/config"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// rpcer is the interface needed by a prevAllocWatcher to make RPC calls.
+type rpcer interface {
+	// RPC allows retrieving remote allocs.
+	RPC(method string, args interface{}, reply interface{}) error
+}
+
+// terminated is the interface needed by a prevAllocWatcher to check if an
+// alloc is terminated.
+type terminated interface {
+	Terminated() bool
+}
+
+// prevAllocWatcher allows AllocRunners to wait for a previous allocation to
+// terminate and migrate its data whether or not the previous allocation is
+// local or remote.
+type prevAllocWatcher interface {
+	// Wait for previous alloc to terminate
+	Wait(context.Context) error
+
+	// Migrate data from previous alloc
+	Migrate(ctx context.Context, dest *allocdir.AllocDir) error
+}
+
+// newAllocWatcher creates a prevAllocWatcher appropriate for whether this
+// alloc's previous allocation was local or remote. If this alloc has no
+// previous alloc then a noop implementation is returned.
+func newAllocWatcher(alloc *structs.Allocation, prevAR *AllocRunner, rpc rpcer, config *config.Config, l *log.Logger) prevAllocWatcher {
+	if alloc.PreviousAllocation == "" {
+		// No previous allocation, use noop transitioner
+		return noopPrevAlloc{}
+	}
+
+	tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
+
+	if prevAR != nil {
```

Comment:
While this conditional is ugly, it is intentionally put inside this unified constructor so that the same logic doesn't have to be duplicated in `client.go`. There's only 2 places that call into this new func, but I'd rather just let them ignore these conditions and leave it up to this file to handle it all.

Definitely up for debate...